src/storage/src/hummock/sstable/writer.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bytes::{BufMut, Bytes, BytesMut};
+use risingwave_hummock_sdk::HummockSstableId;
+
+use crate::hummock::{
+    CachePolicy, HummockResult, SstableBuilderOptions, SstableStoreRef, SstableStreamingUploader,
+};
+
+/// A consumer of SST data.
+/// The data may be written to a stream, pushed to a vector, appended to a buffer, etc.
+#[async_trait::async_trait]
+pub trait SstableWriter: Send {
+    type Output;
+
+    /// Write an SST block to the writer.
+    fn write_block(&mut self, block: &[u8]) -> HummockResult<()>;
+
+    /// Get the length of data that has already been written.
+    fn data_len(&self) -> usize;
+
+    /// Finish writing the SST. Return the output along with final data length.
+    async fn finish(mut self, size_footer: u32) -> HummockResult<(usize, Self::Output)>;
+}
+
+pub struct InMemSstableWriter {
+    buf: BytesMut,
+}
+
+/// Append SST data to a buffer.
+impl InMemSstableWriter {
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            buf: BytesMut::with_capacity(capacity),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl SstableWriter for InMemSstableWriter {
+    type Output = Bytes;
+
+    fn write_block(&mut self, block: &[u8]) -> HummockResult<()> {
+        self.buf.put_slice(block);
+        Ok(())
+    }
+
+    async fn finish(mut self, size_footer: u32) -> HummockResult<(usize, Self::Output)> {
+        self.buf.put_slice(&size_footer.to_le_bytes());
+        Ok((self.buf.len(), self.buf.freeze()))
+    }
+
+    fn data_len(&self) -> usize {
+        self.buf.len()
+    }
+}
+
+impl From<&SstableBuilderOptions> for InMemSstableWriter {
+    fn from(options: &SstableBuilderOptions) -> InMemSstableWriter {
+        InMemSstableWriter::new(options.capacity + options.block_capacity)
+    }
+}
+
+/// Upload sst blocks to the streaming uploader of `SstableStore`.
+/// The upload might not be finished immediately on calling `finish`.
+/// Instead, it returns a sealed uploader which is ready to be passed back to `SstableStore` to
+/// finish the uploading.
+pub struct StreamingSstableWriter {
+    uploader: SstableStreamingUploader,
```

Comment:
There are too many concepts introduced for the multipart streaming upload support:
1. ObjectStreamingUploader
2. SstableStreamingUploader
3. SstableWriter
4. StreamingUploadSealer
5. SstableWriterBuilder

I think we can at least unify 2,3,4 into one thing.

Reply:
3, 4 are concepts on SST builder level, intended to decouple sst building and uploading (consuming, more generally speaking), and 2 is on `SstableStore` level, and I want to separate the concepts of these two levels.

Reply:
I have merged 3 and 4, but 2 and 3,4 are on different levels (2 is on `SstableStore`, 3,4 is on `SstableBuilder`).

Reply:
What is the benefit of separating `SstableStreamingUploader` and `StreamingUploadWriter`? 

They both understand SST/blocks. `SstableBuilder` won't understand the implementation detail of the `SstableWriter/StreamingUploadWriter` anyway and `StreamingUploadWriter` is just a thin wrapper on top of `SstableStreamingUploader`.  Also, `SstableWriterBuilder` always needs to understand SstableStore with or without this wrapper.

I am thinking about whether we should only provide one API in `SstableStore` to write SST. Example:
```
enum SstWriteMode {
  BATCH,
  STRAMING,
}

pub trait SstableStoreWrite: Send + Sync {
    async fn create_sst_writer(
        &self,
        sst_id: HummockSstableId,
        policy: CachePolicy,
        mode: SstWriteMode 
    ) -> HummockResult<Box<dyn SstableWriter>>;
}

pub trait SstableWriter: Send {
    type Output;
    fn write_block(&mut self, block: &[u8]) -> HummockResult<()>;

    fn finish(
        self,
        meta: SstableMeta,
    ) -> HummockResult<Self::Output>;

    fn data_len(&self) -> usize;
}

impl<F> CapacitySplitTableBuilder<F>
where
    F: TableBuilderFactory,
{
    pub fn seal_current(&mut self) -> HummockResult<()> {
        if let Some(builder) = self.current_builder.take() {
            let builder_output = builder.finish()?;
            
            ...

           let writer_output = tokio::spawn(async move {
                let tracker = tracker.map(|mut t| {
                    if !t.try_increase_memory(data.capacity() as u64 + meta.encoded_size() as u64) {
                        tracing::debug!(...);
                    }
                builder_output.writer.finish(meta, self.tracker.take())?
            }); 
            self.sst_outputs.push(SplitTableOutput {
                sst_info,
                writer_output,
                bloom_filter_size,
            });
        }
        Ok(())
    }
```

cc @Gun9niR @Little-Wallace @wenym1 What do you think?

Reply:
The benefit is it hides implmentation detail of `SstableStoreWrite`.

In your purposal, we need another interface on `SstableStoreWrite`: `finish_sst_writer` to prevent leaking implementation details of `SstableStoreWrite` to `SstableWriter`. For example, we cannot simply pass the `lru_cache` and `block_cache` members to the writer, because `sstable_store` is now a dynamic type and we cannot assume these two caches. The only thing we can do is to get the writer back and let the implementation of `SstableStoreWrite` to determine what to do with the data. And there comes the contradiction: we cannot get the data out of `SstableWriter`, because the type of the writer is dynamic due to the merging of two upload modes.

A potential solution would be to add an interface like `add_cache` to `SstableStoreWrite`, but it's very hard to determine the when to add cache or add what to cache.

Reply:
`SstableWriter` is a trait. A concrete implementation of `SstableStoreWrite` (e.g. `SstableStore`) is responsible for creating the concrete `SstableWriter` when `create_sst_writer` is called so the implementation details are hidden and not leaked. We don't need `finish_sst_writer` because we can refill caches in `SstableWriter::finish` if the `SstableWriter` is created by `SstableStore` with block/meta cache.

Reply:
LGTM. @wenym1 @Little-Wallace what do you think?

Reply:
LGTM too.